AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -977,91 +796,134 @@ - (ASCellNode *)nodeAtCompletedIndexPath:(NSIndexPath *)indexPath
+      // Note: Node may not be allocated and laid out yet (e.g node virtualization or same size optimization)
+      // TODO: Force synchronous allocation and layout pass in that case?
+      [nodesInSection addObject:context.node];
+    }
+    [completedNodes addObject:nodesInSection];
   }
+  return completedNodes;
+}
 
-  NSInteger section = 0;
-  // Loop through each section to look for the cellNode
-  NSString *kind = cellNode.supplementaryElementKind ?: ASDataControllerRowNodeKind;
-  for (NSArray *sectionNodes in [self completedNodesOfKind:kind]) {
-    NSUInteger item = [sectionNodes indexOfObjectIdenticalTo:cellNode];
-    if (item != NSNotFound) {
-      return [NSIndexPath indexPathForItem:item inSection:section];
-    }
-    section += 1;
+- (void)moveCompletedNodeAtIndexPath:(NSIndexPath *)indexPath toIndexPath:(NSIndexPath *)newIndexPath
+{
+  ASDisplayNodeAssertMainThread();
+  // TODO Revisit this after native move support. Need to update **both**
+  // _nodeContexts and _completedNodeContexts through a proper threading tunnel.
+  ASMoveElementInTwoDimensionalArray(_nodeContexts[ASDataControllerRowNodeKind], indexPath, newIndexPath);
+}
+
+#pragma mark - External supplementary store and section context querying
+
+- (ASCellNode *)supplementaryNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
+{
+  ASDisplayNodeAssertMainThread();
+  if (kind == nil || indexPath == nil) {
+    return nil;
   }
   
-  return nil;
+  ASIndexedNodeContext *context = ASGetElementInTwoDimensionalArray(_completedNodeContexts[kind], indexPath);
+  // Note: Node may not be allocated and laid out yet (e.g node virtualization or same size optimization)
+  // TODO: Force synchronous allocation and layout pass in that case?
+  return context.node;
 }
 
-/// Returns nodes that can be queried externally. _externalCompletedNodes is used if available, _completedNodes otherwise.
-- (NSArray *)completedNodes
+- (id<ASSectionContext>)contextForSection:(NSInteger)section
 {
   ASDisplayNodeAssertMainThread();
-  return _externalCompletedNodes ? : _completedNodes[ASDataControllerRowNodeKind];
+  ASDisplayNodeAssertTrue(section >= 0 && section < _sections.count);
+  return _sections[section].context;
 }
 
-- (void)moveCompletedNodeAtIndexPath:(NSIndexPath *)indexPath toIndexPath:(NSIndexPath *)newIndexPath
+#pragma mark - Node contexts dictionary
+
+//TODO Move this to somewhere else?
++ (ASNodeContextTwoDimensionalDictionary *)deepImmutableCopyOfNodeContextsDictionary:(ASNodeContextTwoDimensionalDictionary *)originalDict
 {
-  ASDisplayNodeAssertMainThread();
-  ASMoveElementInTwoDimensionalArray(_externalCompletedNodes, indexPath, newIndexPath);
-  ASMoveElementInTwoDimensionalArray(_completedNodes[ASDataControllerRowNodeKind], indexPath, newIndexPath);
+  NSMutableArray<ASNodeContextTwoDimensionalArray *> *deepCopy = [NSMutableArray arrayWithCapacity:originalDict.count];
+  for (ASNodeContextTwoDimensionalArray *sections in originalDict.allValues) {
```

Comment:
Rather than copying the array with `allValues`, let's use `objectEnumerator` here and below.